### DISCUSSION TITLE
Not getting transparency to work

### DISCUSSION THREAD

**Discussion by MichaMeyer** [chillerlan/php-qrcode#196] [Q&A]
Hi Codemasher!

Sukith and his team have unfortunately left me without a solution!

I now decided to omit the fixed size and only use scale(), furthermore I have included your suggestion from here: https://github.com/chillerlan/php-qrcode/discussions/189#discussioncomment-5409546

This is the current state of the scripts. 

## LogoOptions.php
```php
<?php

declare(strict_types=1);

namespace App\QR\Options;

use chillerlan\QRCode\QROptions;

class LogoOptions extends QROptions
{
    // size in QR modules, multiply with QROptions::$scale for pixel size
    protected int $logoSpaceWidth;
    protected int $logoSpaceHeight;

    // additional options for logo
    protected int $logoWidth;
    protected int $logoHeight;

    public function __construct(array $options = [])
    {
        parent::__construct($options);

        $this->logoWidth = $options['logoWidth'] ?? 0;
        $this->logoHeight = $options['logoHeight'] ?? 0;
    }

    public function getLogoWidth(): int
    {
        return $this->logoWidth;
    }

    public function getLogoHeight(): int
    {
        return $this->logoHeight;
    }
} 
```
## separateData.php
( separates Url into param for QR Creation )
```php
<?php

$actual_link = (empty($_SERVER['HTTPS']) ? 'http' : 'https') . "://$_SERVER[HTTP_HOST]$_SERVER[REQUEST_URI]";
$fullUrl= $_SERVER['REQUEST_URI'];
$dataValueParts = explode('/', $fullUrl);
$url = $dataValueParts[1];

$parts = explode('-', $fullUrl);
$imageSize = explode('x', $parts[2]);

$imageDownloadName = $fullUrl;
$imageDownloadName = substr($imageDownloadName, 1);
$imageDownloadNameEncoded= urlencode($imageDownloadName);

$widthValue = intval($imageSize[0]);
$heightValue = intval($imageSize[1]);
$colorValueDot = '#' . $parts[3];
$colorValueBack = '#' . $parts[4];

$username = $parts[6];
$prefix_url = 'https://devel.domain.com/'.$username;
$downExt = $parts[1];
$logo = $parts[5];

$colorValueDot = hexToRgb($colorValueDot);
$colorValueBack = hexToRgb($colorValueBack);

if (strval($parts[4])== 'T')
{

	$colorValueBack = [0, 250, 0];

}


function hexToRgb($hex)
{
    $hex = str_replace("#", "", $hex); // remove the # if it exists
    if (strlen($hex) == 3) { // if shorthand notation is used
      $hex = str_repeat(substr($hex, 0, 1), 2).str_repeat(substr($hex, 1, 1), 2).str_repeat(substr($hex, 2, 1), 2);
    }
    $rgb = array(); // initialize an empty array for the RGB values
    $rgb[] = hexdec(substr($hex, 0, 2)); // convert the red hex value to decimal and add it to the array
    $rgb[] = hexdec(substr($hex, 2, 2)); // convert the green hex value to decimal and add it to the array
    $rgb[] = hexdec(substr($hex, 4, 2)); // convert the blue hex value to decimal and add it to the array
    return $rgb; // return the array of RGB values
}
```
## index.php
```php
<?php

  declare(strict_types=1);

  use chillerlan\QRCode\QRCode;
  use chillerlan\QRCode\QROptions;
  use App\QR\Image\QRImageWithLogo;
  use App\QR\Options\LogoOptions;
  use chillerlan\QRCode\Data\QRMatrix;

  require_once('./../vendor/autoload.php');
  include 'separateData.php';

  $data = $prefix_url;

  $options = new LogoOptions(
    [
      'eccLevel' => QRCode::ECC_H,
      'imageBase64' => true,
      'logoSpaceHeight' => 13,
      'logoSpaceWidth' => 13,
      'version' => 4,
	  'scale' => 60,
	  'quietzoneSize' => 2,
	  'imageTransparent'=> true,
	  'transparencyColor' => [0, 250, 0],
      'outputType'   => QRCode::OUTPUT_IMAGE_PNG,
      'moduleValues' => [
		// finder
		1536 => $colorValueDot, //need
		6    => $colorValueBack,
     // no  // light (false), white is the transparency color and is enabled by default
		5632 => $colorValueDot,// yes finder dot, dark (true)
		// alignment
		2560 => $colorValueDot, //yes    // [201, 229, 225]
		10   => $colorValueBack, // no
		// timing
		3072 => $colorValueDot, //yes
		12   => $colorValueBack, // no
		// format
		3584 => $colorValueDot, //yes
		14   => $colorValueBack, //no
		// version
		4096 => $colorValueDot, //yes
		16   => $colorValueBack,
		// data
		1024 => $colorValueDot, //yes
    	4    => $colorValueBack,
		// darkmodule
		512  => $colorValueDot, //yes,
		// separator
		8    => $colorValueBack,
		// quietzone
		18   => $colorValueBack,//b order
		// logo (requires a call to QRMatrix::setLogoSpace())
		20    => $colorValueBack, //logo border
	],
    ]
  );

  $qrOutputInterface = new QRImageWithLogo(
    $options,
    (new QRCode($options))->getMatrix($data)
  );

  $qrCode = $qrOutputInterface->dump(
    null,
    __DIR__.'/../public/img/'.$logo.'.png'
  );

// Remove the data URI scheme and get the raw image data
$imageData = base64_decode(preg_replace('#^data:image/\w+;base64,#i', '', $qrCode));

// Create an image resource from the raw image data
$image = imagecreatefromstring($imageData);

// Specify the custom dimensions
$newWidth = 500;
$newHeight = 500;


//$trans_bg_color = imagecolorat($image,1,1);
//$trans_bg_color = imagecolorallocate($image, 0, 0, 0);
//imagecolortransparent($image, $trans_bg_color);

// Create a new image with the custom dimensions
//$newImage = imagecreatetruecolor($newWidth, $newHeight);

// Resize and copy the original image to the new image with the custom dimensions
//imagecopyresampled($newImage, $image, 0, 0, 0, 0, $newWidth, $newHeight, imagesx($image), imagesy($image));

// Specify the file path to save the image
$filePath = __DIR__ .'/../public/cache/'.$imageDownloadNameEncoded.'.png';

// Save the image to the file path
//imagepng($image, $filePath);

// Get the MIME type of the image
$mimeType = 'image/png'; // You can change this to 'image/jpeg' if the image is a JPEG

// Output the image to the browser
header('Content-Type: ' . $mimeType);
imagepng($image);

// Free up memory by destroying the image resource
imagedestroy($image);

?>
```
so if I use "000000" as foregroundcolor and "T" as indicator for transparent background (which creates $colorValueBack = [0, 250, 0]).

Nevertheless, the output is still without a transparency in the background.

I get following result:
![grafik](https://user-images.githubusercontent.com/32386846/234237368-2bdef100-480a-4898-9673-eaac6487a0a5.png)

 What is going wrong?

thanks a lot for your help! I assume it only could be a small problem.

Another question: is there a possibility to store a "colour profile" in the finished PNG?

BR Rene

**Comment by codemasher** [comment]
I said it before and I'll say it again: this is not an issue of the library but a general issue with PHP's GD image processing extension and how it handles transparency internally. This is beyond my knowledge and I'd suggest you check out [StackOverflow](https://stackoverflow.com/search?q=[php]+[gd]+transparency) or [r/PHPhelp on reddit](https://www.reddit.com/r/PHPhelp/). I can help you with implementation details of my own library, such as doing all of that in a single class so that you don't have the mess all over your scripts. What i can't (or won't) do is help you to figure out how transparency is properly retained as it would require me to do the same research work.
Also, i already suggested a solution that *might* work but you have to try it for yourself: https://www.php.net/manual/en/function.imagecolortransparent.php#77099

**Reply by codemasher** [reply]
btw this is the first result of the StackOverflow query i linked: https://stackoverflow.com/questions/32243/can-png-image-transparency-be-preserved-when-using-phps-gdlib-imagecopyresample

**Reply by MichaMeyer** [reply]
Sorry codemasher! It seems that there is a misunderstanding! 

We have completely changed definition! **we do NOT want to scale the PNG anymore** to prevent the scaling problem!
Removed all "external" processing so that we could not have any PNG issue? 

this "mess" was created from my "professional developers" from Sri Lanka, we have ended cooperation because they did not get it to work.
I'm sorry about this!

Regarding "_implementation details of my own library, such as doing all of that in a single class_"
I would really appreciate your help with only your library!

We just want to use **the logic from your library**, nothing more.

1. Creating QR Code with transparent background using internal scale () function (Transpareny feature from your lib?)
2. Adding logo in the middle with logospace
3. Exporting as PNG with GDLIB (is also in your examples) 
4. **No scaling or resizing AFTER QR Code creation what was the problem earlier.**

Just support for your functions in your library? I can also make some donation via Paypal if you can help? 
NO GDLIB Support or anything else.

I would have not asked for the same thing as before, please don't think this? I have understood that the problems occur when scaling PNG afterwards? the example I posted is without any scaling?

We decided to only use the features your library supports? just to make it work for us!

Or does transparency feature in general does not work anymore in the library?

**Reply by codemasher** [reply]
I should clarify that extending the library as shown in the examples with functionality beyond my knowledge still requires *you* to do the research about how things work - I am not general PHP support. Which means: you need figure out how the thing you want to do (add logo, scale, retain transparency for your use case), I can help connecting that with the library's internals. The examples are mere proof-of-concepts as an idea to start from. They're never meant to be complete or functional under any given circumstance.

**Reply by MichaMeyer** [reply]
Good Morning!
Yes, that's completely clear! As it seems, the only thing left not working is the transparency feature. All other things work as expected.

Are **'imageTransparent'=> true**, 
**'transparencyColor' => [0, 250, 0]** and 
**$colorValueBack = [0, 250, 0];** 

correctly used and handled to get a transparent background? Or what needs to be different?
Thanks a lot!

**Reply by codemasher** [reply]
Hey,

> 'imageTransparent'=> true,
> 'transparencyColor' => [0, 250, 0] and
> $colorValueBack = [0, 250, 0];

There seems nothing wrong with these values. This sets the "light" color to green and makes this the color that's turned transparent by `imagecolortransparent()`. It should remain transparent after output as long as you don't put it into `imagecreatefromstring()` or similar after that.

**Reply by MichaMeyer** [reply]
Hi Codemasher!

1. I removed all the scaling code and changed  'imageBase64' => true, to" false" and so I was able to remove  "base64_decode(preg_replace" function. 

  So I have the "Logo Options" 
  then
  $qrOutputInterface = new QRImageWithLogo($options,(new QRCode($options))->getMatrix($data));
  $qrCode = $qrOutputInterface->dump(null,__DIR__.'/../public/img/'.$logo.'.png');
  $image = imagecreatefromstring($imageData);
  $mimeType = 'image/png';
  header('Content-Type: ' . $mimeType);
  imagepng($image);
  imagedestroy($image);

Then I tried to get rid of imagecreatefromstring() 

I tried several output variants from your examples, but this does not work without errors, so I also deactivated any transparency for testing.
I tried also
**Direct output of matrix** 
or 
**the example from ImageWithLogo.php where I changed:** 

$qrCode = $qrOutputInterface->dump(null,__DIR__.'/../public/img/'.$logo.'.png');
to 
echo $qrOutputInterface->dump(null,__DIR__.'/../public/img/'.$logo.'.png');

but I only get something like "**�PNG  IHDR � �g}ѣtRNS���7X} pHYs���**" in browser

I tried alternatives like 
    $image = imagecreatefrompng('data://image/png;base64,'.$qrcode)
    imagepng($image);

as this is suggested **on stack overflow**, but this also did not work!

Do you have a suggestion, how I can create a qrcode **based on existing logic above** WITHOUT the imagecreatefromstring() function? Nothing special in first step, just a working PNG.

Thank you!

**Reply by codemasher** [reply]
This is now like the 5th time you're askin that and i suggested several times before that you do all of your modification **within** your custom output class. or use the `returnResource` option so that you get the GD image resource. Alternatively try your luck with the ImagMagick output.
